nutaudet/MTCDetector.cxx

```diff
@@ -0,0 +1,543 @@
+    //   // Place the SciFi module (fiber module)
+    //   envVol->AddNode(sciFiModuleVol, i, new TGeoTranslation(0, 0, zPos - fScintThick/2));
+    //   // Place the Scintillator layer (shifted up by half the iron thickness)
+    //   envVol->AddNode(scintVol, i, new TGeoTranslation(0, 0, zPos + fIronThick/2));
+
+    // --- Assemble the layers into the envelope ---
+    for (Int_t i = 0; i < fLayers; i++) {
+        // Compute the center position (z) for the current module
+        Double_t zPos = -totalLength / 2 + i * moduleSpacing;
+
+        // Place the Outer Iron layer (shifted down by half the SciFi+scint thickness)
+        envVol->AddNode(ironVol, i, new TGeoTranslation(0, 0, zPos + fIronThick / 2));
+
+        // Create a SciFi module with the current detector id 'i'
+        TGeoVolume* sciFiModuleVol = CreateSciFiModule("MTC_sciFi", fWidth, fHeight, fSciFiThick, i);
+        envVol->AddNode(sciFiModuleVol, i, new TGeoTranslation(0, 0, zPos + fIronThick + fSciFiThick / 2));
+        TGeoVolume* scintVol =
+            CreateSegmentedLayer("MTC_scint", fWidth, fHeight, fScintThick, 1.0, 1.0, scintMed, kAzure + 7, 30, i);
+        // Place the Scintillator layer (shifted up by half the iron thickness)
+        envVol->AddNode(scintVol, i, new TGeoTranslation(0, 0, zPos + fIronThick + fSciFiThick + fScintThick / 2));
+    }
+
+    // Finally, add the envelope to the top volume with the global z offset fZCenter
+    gGeoManager->GetTopVolume()->AddNode(envVol, 1, new TGeoTranslation(0, 0, fZCenter));
+}
+// Standard FairDetector methods
+void MTCDetector::Initialize()
+{
+    FairDetector::Initialize();
+}
+
+Bool_t MTCDetector::ProcessHits(FairVolume* vol)
+{
+    /** This method is called from the MC stepping */
+    // Set parameters at entrance of volume. Reset ELoss.
+    if (gMC->IsTrackEntering()) {
+        fELoss = 0.;
+        fTime = gMC->TrackTime() * 1.0e09;
+        fLength = gMC->TrackLength();
+        gMC->TrackPosition(fPos);
+        gMC->TrackMomentum(fMom);
+    }
+    // Sum energy loss for all steps in the active volume
+    fELoss += gMC->Edep();
+
+    // Create vetoPoint when exiting active volume
+    if (gMC->IsTrackExiting() || gMC->IsTrackStop() || gMC->IsTrackDisappeared()) {
+
+        if (fELoss == 0.) {
+            return kFALSE;
+        }   // if you do not want hits with zero eloss
+
+        TParticle* p = gMC->GetStack()->GetCurrentTrack();
+        fTrackID = gMC->GetStack()->GetCurrentTrackNumber();
+        Int_t pdgCode = p->GetPdgCode();
+        Int_t detID;
+        gMC->CurrentVolID(detID);
+        TLorentzVector Pos;
+        gMC->TrackPosition(Pos);
+        TLorentzVector Mom;
+        gMC->TrackMomentum(Mom);
+        Double_t xmean = (fPos.X() + Pos.X()) / 2.;
+        Double_t ymean = (fPos.Y() + Pos.Y()) / 2.;
+        Double_t zmean = (fPos.Z() + Pos.Z()) / 2.;
+        // cout << "Check detID: " << detID << endl;
+        if (detID / 10000000 == 3) {
```

Comment:
It might be better to just define the position vector conditionally to avoid duplicating everything.